userspace/libscap/scap_open.h

```diff
@@ -31,79 +31,125 @@ limitations under the License.
+		 * returned.
+		 */
+		SCAP_MODE_NODRIVER,
+		/*!
+		 * Do not read system call data. Events come from the configured input plugin.
+		 */
+		SCAP_MODE_PLUGIN,
+	} scap_mode_t;
+
 	/*!
-	 * Read system call data from the underlying operating system using a modern
-	 * bpf probe.
+	 * \brief Argument for scap_open
+	 * Set any PPM_SC syscall idx to true to enable its tracing at driver level,
+	 * otherwise syscalls are not traced (so called "uninteresting syscalls").
 	 */
-	SCAP_MODE_MODERN_BPF,
-#endif
-} scap_mode_t;
-
-/*!
-  \brief Argument for scap_open
-  Set any PPM_SC syscall idx to true to enable its tracing at driver level,
-  otherwise syscalls are not traced (so called "uninteresting syscalls").
-*/
-typedef struct {
-	bool ppm_sc[PPM_SC_MAX];
-} interesting_ppm_sc_set;
-
-typedef struct scap_open_args
-{
-	scap_mode_t mode;
-	int fd; // If non-zero, will be used instead of fname.
-	const char* fname; ///< The name of the file to open. NULL for live captures.
-	proc_entry_callback proc_callback; ///< Callback to be invoked for each thread/fd that is extracted from /proc, or NULL if no callback is needed.
-	void* proc_callback_context; ///< Opaque pointer that will be included in the calls to proc_callback. Ignored if proc_callback is NULL.
-	bool import_users; ///< true if the user list should be created when opening the capture.
-	uint64_t start_offset; ///< Used to start reading a capture file from an arbitrary offset. This is leveraged when opening merged files.
-	const char *bpf_probe; ///< The name of the BPF probe to open. If NULL, the kernel driver will be used.
-	const char *suppressed_comms[SCAP_MAX_SUPPRESSED_COMMS]; ///< A list of processes (comm) for which no
-	// events should be returned, with a trailing NULL value.
-	// You can provide additional comm
-	// values via scap_suppress_events_comm().
-	bool udig; ///< If true, UDIG will be used for event capture.
-	bool gvisor; //< If true, gVisor will be used for event capture
-	const char *gvisor_root_path; ///< When using gvisor, the root path used by runsc commands
-	const char *gvisor_config_path; ///< When using gvisor, the path to the configuration file
-
-	interesting_ppm_sc_set ppm_sc_of_interest;
-
-	scap_source_plugin* input_plugin; ///< use this to configure a source plugin that will produce the events for this capture
-	char* input_plugin_params; ///< optional parameters string for the source plugin pointed by src_plugin
-
-	scap_test_input_data* test_input_data; ///< only used for testing scap consumers by supplying arbitrary test data
-}scap_open_args;
+	typedef struct
+	{
+		bool ppm_sc[PPM_SC_MAX];
+	} interesting_ppm_sc_set;
+
+	typedef enum
+	{
+		UNKNOWN_ENGINE = 0,
+		KMOD_ENGINE = 1,
+		BPF_ENGINE = 2,
+		UDIG_ENGINE = 3,
+		NODRIVER_ENGINE = 4,
+		SAVEFILE_ENGINE = 5,
+		PLUGIN_ENGINE = 6,
+		GVISOR_ENGINE = 7,
+		MODERN_BPF_ENGINE = 8,
+		TEST_INPUT_ENGINE = 9,
+	} scap_engine_t;
```

Comment:
Probably an enum is clearer and since we don't add a new engine every day, it could be a reasonable price to pay, but I have not a strong opinion on that, if you have already implemented all your changes based on strings I can change this enum in strings or you could also push on this branch if you want :) 
I imagine you have used something like 

```c
char engine_name[MAX_NAME_LEN]; 
```

Plus some defines like `#define KMOD_ENGINE kmod`. Is that true?

Reply:
I just used `const char*` with compile-time constant strings 🤷 

I know neither an enum nor a string is ideal. My goal would be to make the machinery flexible enough that you can add engines without rebuilding libscap itself (or develop a new engine without touching anything besides libscap/engines/foo and a CMakeLists.txt) but just about anything will be an improvement over the current state.